File: web/src/model/victron_controller/dashboard/ZappiDrainBranch.ts
// @ts-nocheck
import {BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'

export enum ZappiDrainBranch {
    Tighten = "Tighten",
    Relax = "Relax",
    Bypass = "Bypass",
    Disabled = "Disabled"
}

export const ZappiDrainBranch_values: ReadonlyArray<ZappiDrainBranch> = [
    ZappiDrainBranch.Tighten,
    ZappiDrainBranch.Relax,
    ZappiDrainBranch.Bypass,
    ZappiDrainBranch.Disabled
] as const;

export function ZappiDrainBranch_parse(s: string): ZappiDrainBranch {
    const found = ZappiDrainBranch_values.find(v => v === s);
    if (found === undefined) {
        throw new Error("Unknown ZappiDrainBranch variant: " + s);
    }
    return found;
}

export class ZappiDrainBranch_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: ZappiDrainBranch, writer: BaboonBinWriter): unknown {
        if (this !== ZappiDrainBranch_UEBACodec.lazyInstance.value) {
          return ZappiDrainBranch_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        switch (value) {
            case "Tighten": BinTools.writeByte(writer, 0); break;
                case "Relax": BinTools.writeByte(writer, 1); break;
                case "Bypass": BinTools.writeByte(writer, 2); break;
                case "Disabled": BinTools.writeByte(writer, 3); break;
            default: throw new Error("Unknown enum variant: " + value);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): ZappiDrainBranch {
        if (this !== ZappiDrainBranch_UEBACodec .lazyInstance.value) {
            return ZappiDrainBranch_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return "Tighten" as ZappiDrainBranch;
                case 1: return "Relax" as ZappiDrainBranch;
                case 2: return "Bypass" as ZappiDrainBranch;
                case 3: return "Disabled" as ZappiDrainBranch;
            default: throw new Error("Unknown enum variant tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return ZappiDrainBranch_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return ZappiDrainBranch_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#ZappiDrainBranch'
    public baboonTypeIdentifier() {
        return ZappiDrainBranch_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new ZappiDrainBranch_UEBACodec())
    public static get instance(): ZappiDrainBranch_UEBACodec {
        return ZappiDrainBranch_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/zappi_drain_branch.rs


#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub enum ZappiDrainBranch {
    Tighten,
    Relax,
    Bypass,
    Disabled,
}

impl ZappiDrainBranch {
    pub fn parse(s: &str) -> Result<Self, String> {
        match s {
            "Tighten" => Ok(ZappiDrainBranch::Tighten),
            "Relax" => Ok(ZappiDrainBranch::Relax),
            "Bypass" => Ok(ZappiDrainBranch::Bypass),
            "Disabled" => Ok(ZappiDrainBranch::Disabled),
            _ => Err(format!("Unknown variant: {}", s)),
        }
    }

    pub fn all() -> Vec<ZappiDrainBranch> {
        vec![
            ZappiDrainBranch::Tighten,
            ZappiDrainBranch::Relax,
            ZappiDrainBranch::Bypass,
            ZappiDrainBranch::Disabled,
        ]
    }
}

impl std::fmt::Display for ZappiDrainBranch {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ZappiDrainBranch::Tighten => write!(f, "Tighten"),
            ZappiDrainBranch::Relax => write!(f, "Relax"),
            ZappiDrainBranch::Bypass => write!(f, "Bypass"),
            ZappiDrainBranch::Disabled => write!(f, "Disabled"),
        }
    }
}

impl std::str::FromStr for ZappiDrainBranch {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse(s)
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for ZappiDrainBranch {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for ZappiDrainBranch {
    fn encode_ueba(&self, _ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            ZappiDrainBranch::Tighten => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
            ZappiDrainBranch::Relax => crate::baboon_runtime::bin_tools::write_byte(writer, 1)?,
            ZappiDrainBranch::Bypass => crate::baboon_runtime::bin_tools::write_byte(writer, 2)?,
            ZappiDrainBranch::Disabled => crate::baboon_runtime::bin_tools::write_byte(writer, 3)?,
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for ZappiDrainBranch {
    fn decode_ueba(_ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => Ok(ZappiDrainBranch::Tighten),
            1 => Ok(ZappiDrainBranch::Relax),
            2 => Ok(ZappiDrainBranch::Bypass),
            3 => Ok(ZappiDrainBranch::Disabled),
            _ => Err(format!("Unknown enum variant tag: {}", tag).into()),
        }
    }
}